design/airgap-existing-cluster.md

```diff
@@ -0,0 +1,109 @@
+
+## Goals
+
+- Install kotsadm to an existing cluster where the cluster does not have internet access
+- Install kotsadm to an existing cluster from a workstation where the workstation does not have internet access at the time of installation
+- Support different credentials for push / pull (workstation / cluster) to the local registry
+- Support temporary/dynamic and persistent/static credentials for pushing kotsadm images to the registry
+- Support only persistent/static credentials for pulling kotsadm images to the registry
+- Support different credentials for pushing vs pulling images to the local registry
+
+## Non Goals
+
+- Create a single installer for kotsadm + application
+- Create a single download for kotsadm, kots plugin, application and license
+- Automated installs of kotsadm to existing clusters in airgap mode
+- Support for installing versions of KOTS different from the KOTS CLI version installed
+
+## Background
+
+It's not currently possible to install KOTS Admin Console into an existing cluster that doesn't have outbound internet access.
+
+## High-Level Design
+
+Airgap Workstation: the computer with kubectl access to the airgap cluster
+Public Workstation: the computer with access to the internet
+
+From the public workstation:
+1. The user downloads the kots plugin suitable for Airgap Workstation.
+1. The user downloads the kotsadm archive containing kots container images
+1. The user downloads the airgap bundle of the application
+1. The user copies the kots plugin, the kotsadm archive and the application airgap bundle onto a medium (eg a USB stick) that can be transfered into the airgapped network.
+
+From the airgap workstation:
+1. The user will install the kots plugin 
+1. The user will run a `kubectl kots` command to copy the container images in the kotsadm archive to the local registry, providing registry details for pushing
+1. The user will install kotsadm to the cluster using the kots plugin via `kots install` passing local registry information which will be used to generate the kotsadm yaml and image pull secrets (if neccessary)
+1. The user will install the application and license using the existing airgap installation process 
+
+## Detailed Design
+
+The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
+The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.
+The files SHOULD be stored in the archive in the same format as application airgap files are stored.
+
+Add a `kots admin-console push-images` to the kots cli to retag and push all kotsadm images to an authenticated registry that is accessible from the cluster, using the desired push credentials.
+This new command is used to copy all of the KOTS Admin Console images from one location to another.
+
+Example:
+
+```
+kubectl kots admin-console push-images ./kotsadm-1.16.1.tar.gz registry.somebigbank.com/my-app \
+  --registry-username push-user \
```

Comment:
I propose we replace `registry-user` and `registry-password` with a `registry-secret-name`.   The secret will contain a `.dockerconfigjson` key and the json will contain a key for every registry used by the process (the push registry and the pull registry).  This is the standard k8s way to define registry secrets and this gets us automation at no extra cost.

Reply:
Talked with @divolgin and we can skip this for now. If we hit the stretch goal above of trying to pull anonymously and if rejected then try with `~/.docker/config.json` credentials, then someone could `docker login` before hand and this would work without leaking credentials into bash history.